File: Java/Generics/IterationTest/SimpleStack.java
import java.util.Iterator;

class SimpleStack<V> implements Iterable<V> {

	class Node {

		V value;
		Node below;

		Node(V val) {
			value = val;
			below = top;
		}
	}

	private Node top;

	public void push(V item) {
		top = new Node(item);
	}

	public V pop() {
		V item = top.value;
		top = top.below;
		return item;
	}

	public boolean empty() {
		return top == null;
	}

	public Iterator<V> iterator() {
		//creating a new instance of an inner local anonymous class
		//which implements java.util.Iterator interface
		return new Iterator<V>() {

			Node current = top;

			public boolean hasNext() {
				return current != null;
			}

			public V next() {
				V result = current.value;
				current = current.below;
				return result;
			}
		};
	}
}

File: DotNet/Generics/IterationTest/DemoApp/SimpleStack.cs
class SimpleStack<V>
{
    class Node
    {
        internal V Value;
        internal Node Below;

        internal Node(V val, Node bel)
        {
            Value = val;
            Below = bel;
        }
    }

    private Node top;

    public void Push(V item)
    {
        top = new Node(item, top);
    }

    public V Pop()
    {
        V item = top.Value;
        top = top.Below;
        return item;
    }

    public bool Empty()
    {
        return top is null;
    }

    public IEnumerator<V> GetEnumerator()
    {
        for(Node n = top; n != null; n = n.Below)
            yield return n.Value; 
            //compiler will generate implementation for IEnumerator<V> for yielded values
    }
}
